### PR TITLE
[makeotf] Use correct lookup type for `ignore rsub`

### PR DESCRIPTION
**Pull request by khaledhosny**
The code was unconditionally overriding the lookup type to become Chaining Contextual Substitution, instead do it only if it is not Reverse Chaining Contextual Substitution.

Fixes https://github.com/adobe-type-tools/afdko/issues/992

### CODE REVIEW

**Comment by khaledhosny** [comment]
While doing some cleanup to the code, I also fixed https://github.com/adobe-type-tools/afdko/issues/1005 by dropping the nonworking and undocumented special support for `NULL` substitution.

File: tests/makeotfexe_data/input/spec/5fii-4.fea [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Test comma-separated ingore sub statements.
```
Comment (punchcutter) [inline tests/makeotfexe_data/input/spec/5fii-4.fea:1]: typo _ignore_

File: tests/makeotfexe_data/input/spec/5h-6.fea [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Test comma-separated ingore rsub statements.
```
Comment (punchcutter) [inline tests/makeotfexe_data/input/spec/5h-6.fea:1]: typo _ignore_

**Comment by khaledhosny** [comment]
I didn’t actually do any testing, I just verified that the generated GSUB table look as expected (a Reverse Substitution Lookup with no substitutions. The semantics of the text file may be nonsense, I’ll need to heck this and see what should be the expected behavior.

**Comment by khaledhosny** [comment]
Here is my finding so far, and `ReverseChainSingleSubstFormat1` subtable with empty `substituteGlyphIDs` will be ignore by Uniscribe and Core Text, but HarfBuzz will use GID 0 (`.notdef`) for substitute glyph (looks like a bug), so I guess empty substitutes is an unsupported case (unlike chaining substitution subtables, reverse chaining does not reference other lookups but includes the substitutes directly which might explain the difference).

This might be the reason `ignore rsub` was treated as if it was `ignore sub`, but doing this makes it useless as it would do nothing (as we end with two lookups, not one lookup with two subtables).

I think our options either:
1. disallow `ignore rsub` and make it an error, or
2. make `ignore rsub A' B;` translate to `rsub A' B by A;` (i.e. replace A by itself), which seems to give the desired effect in HarfBuzz and Uniscribe but ignored by Core Text, or
3. Keep things as they are now (broken) since ~no one is complaining.

**Comment by khaledhosny** [comment]
Since `ignore rsub` apparently never worked, and making it work is a bit non-trivial and might be surprising since it is not the same as `ignore sub`, I’m leaning towards option 1). The same effect can be achieved by `rsub A' by A;` which is more explicit and if it cause any surprises they should be more visible to the user.

**Comment by punchcutter** [comment]
I wonder how many people are using rsub in the first place. @khaledhosny are you using it?

**Comment by khaledhosny** [comment]
I’m using `rsub` in one font, yes, and it is working fine. `ignore rsub` is the broken one, and I think it was an untested feature under the mistaken believe that reverse chaining substitution is similar in structure to chaining substitution.

My suggestion would to drop `ignore rsub` from the spec and the code since it never worked and can’t be made to work.

**Comment by punchcutter** [comment]
Yes, I know we are talking about `ignore rsub`, but it goes together with `rsub` so I'm wondering how common it is. Do you have an example of how you are using it that it's working fine?

**Comment by khaledhosny** [comment]
`rsub` is rather rarely used, I’ve only used it once but the project is not publicly available right now. But here is the part using `rsub`. 

```fea
lookup BehAlefOrLam {
  sub [behDotless-ar.init behDotless-ar.medi]'
      [lam-ar.medi]
   by [behDotless-ar.init.prep behDotless-ar.medi.prep];
} BehAlefOrLam;

lookup BehAlefOrLamR {
 rsub [behDotless-ar.init behDotless-ar.medi]'
      [behDotless-ar.init.prep behDotless-ar.medi.prep]
   by [behDotless-ar.init.prep behDotless-ar.medi.prep];
} BehAlefOrLamR;
```
Which basically changes medial `beh` to a unslanted alternate when followed by `lam` (the `sub`), and changes any `beh` before the alternate one to a similar form (the `rsub` part) for any arbitrary sequence of `beh`s (since `rsub` works on the glyph sequence from last to first).

image

image

**Comment by josh-hadley** [comment]
Closing, per discussion. See #1225.

**Pull request closed without merging**